### PR TITLE
Clean up some static functions in staking protocol

### LINKED ISSUE #2142: Clean up some static functions in staking protocol
**What would you like to be added**:
We have some static functions like getTotalBucketCount (not used anywhere), getBucket, updateBucket, putBucket, delBucket, getAllBuckets, getBucketsWithIndices, putBucketAndIndex, which could all be moved in to CandidateStateManager. Also, function settleAction could be moved into CandidateStateManager as well.

**Why is this needed**:
Because of these functions, we have a mixture usage of CandidateStateManager and StateManager throughout this package. By moving them into CandidateStateManager, we don't need to expose StateManager APIs, and CandidateStateManager could be the only interface interacting with this protocol internally.

### PR DESCRIPTION
**Pull request by LuckyPigeon**
# Description

**summary of changes**
Clean up state functions in `bucket_index.go`,  `vote_bucket.go` & `read_state.go`, move them into `CandidateStateManager` and `CandidateStateReader`. Detail changes are shown as below.

**relevant motivation**
We have a mixture usage of CandidateStateManager and StateManager throughout this package. By moving them into CandidateStateManager, we don't need to expose StateManager APIs, and CandidateStateManager could be the only interface interacting with this protocol internally.

**Detail changes**

**CandidateStateReader**

**from `bucket_index.go`**
* getBucketIndices

**from `vote_bucket.go`**
* getBucket
* getTotalBucketCount
* getAllBuckets
* getBucketsWithIndices
* addrKeyWithPrefix
* getCandidate
* getAllCandidates

**from `read_state.go`**
* readStateBuckets
* readStateBucketsByVoter
* readStateBucketsByCandidate
* readStateBucketByIndices
* readStateBucketCount
* readStateCandidates
* readStateCandidateByName
* readStateCandidateByAddress
* readStateTotalStakingAmount

**CandidateStateManager**

**from `bucket_index.go`**
* putBucketIndex
* delBucketIndex

**from `vote_index.go`**
* updateBucket
* putBucket
* delBucket
* putBucketAndIndex
* delBucketAndIndex
* addrKeyWithPrefix
* delCandidate
* putCandidate

Fixes #2142 

## Type of change

Please delete options that are not relevant.

* Bug fix (non-breaking change which fixes an issue)
* New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

**bucket_index_test.go**
* TestBucketIndices

**handler_test.go**
* TestProtocol_HandleCreateStake
* TestProtocol_HandleCandidateRegister
* TestProtocol_HandleCandidateUpdate
* TestProtocol_HandleUnstake
* TestProtocol_HandleWithdrawStake
* TestProtocol_HandleChangeCandidate
* TestProtocol_HandleTransferStake
* TestProtocol_HandleConsignmentTransfer
* TestProtocol_HandleRestake
* TestProtocol_HandleDepositToStake

**candidate_statereader_test.go**

**candidate_statemanager_test.go**

**Test Configuration**:

- Firmware version: Ubuntu 20.04 
- Hardware:
- Toolchain: go1.17.6
- SDK:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by LuckyPigeon** [comment]
@Liuhaai 
Could you help check it failed, it seems `require.Panics` doesn't work properly. A Panic did mount, but `require.Panics` didn't assert successfully.

File: action/protocol/staking/bucket_index.go [hunk 1]
```diff
@@ -68,80 +66,28 @@ func (bis *BucketIndices) deleteBucketIndex(index uint64) {
...
-			&bis,
-			protocol.NamespaceOption(StakingNameSpace),
-			protocol.KeyOption(key))
-	}
-	return err
-}
-
-func getVoterBucketIndices(sr protocol.StateReader, addr address.Address) (*BucketIndices, uint64, error) {
-	return getBucketIndices(sr, addrKeyWithPrefix(addr, _voterIndex))
+func getVoterBucketIndices(csr CandidateStateReader, addr address.Address) (*BucketIndices, uint64, error) {
```
Comment (CoderZhi) [inline action/protocol/staking/bucket_index.go:69]: take this function for example, it is used in function `readStateBucketsByVoter`, which is used in `action/protocol/staking/protocol.go:438`. You may find `csr` has been constructed in line 414, so we may replace line 438 with `csr.BucketsByVoter(ctx, r.GetBucketsByVoter().GetVoterAddress())`. such that we can delete multiple static functions and wrap the logic in csr

File: action/protocol/staking/bucket_index.go [hunk 2]
```diff
@@ -68,80 +66,28 @@ func (bis *BucketIndices) deleteBucketIndex(index uint64) {
...
+	return csr.getBucketIndices(addrKeyWithPrefix(addr, _voterIndex))
```
Comment (CoderZhi) [inline action/protocol/staking/bucket_index.go:70]: `addrKeyWithPrefix(addr, _voterIndex)` could be hidden in csr, thus the caller only needs to provide the voter address to csr

**Comment by LuckyPigeon** [comment]
All state functions in staking protocol are transferred to `CandidateStateManager` and `CandidateStateReader`.

File: action/protocol/staking/bucket_index.go [hunk 3]
```diff
@@ -68,86 +66,26 @@ func (bis *BucketIndices) deleteBucketIndex(index uint64) {
...
-	return err
-}
-
-func getVoterBucketIndices(sr protocol.StateReader, addr address.Address) (*BucketIndices, uint64, error) {
-	return getBucketIndices(sr, addrKeyWithPrefix(addr, _voterIndex))
-}
-
-func putVoterBucketIndex(sm protocol.StateManager, addr address.Address, index uint64) error {
-	return putBucketIndex(sm, addrKeyWithPrefix(addr, _voterIndex), index)
+func getVoterBucketIndices(csr CandidateStateReader, addr address.Address) (*BucketIndices, uint64, error) {
```
Comment (CoderZhi) [inline action/protocol/staking/bucket_index.go:69]: one step further: remove function getVoterBucketIndices by defining two functions:
`func (csr CandidateStateReader) VoterBucketIndices(addr address.Address) (*BucketIndices, uint64, error)`
and 
`func (csr CandidateStateReader) CandidateBucketIndices(addr address.Address) (*BucketIndices, uint64, error)`
Same for the others.

Comment (LuckyPigeon) [reply action/protocol/staking/bucket_index.go]: `VoterBucketIndex` has duplilcated on `del` and `put` use case, so I change to `DelVoterBucketIndex` and `PutVoterBucketIndex`, same for the rest.

File: action/protocol/staking/bucket_index.go [hunk 4]
```diff
@@ -68,86 +66,26 @@ func (bis *BucketIndices) deleteBucketIndex(index uint64) {
...
-	return err
-}
-
-func getVoterBucketIndices(sr protocol.StateReader, addr address.Address) (*BucketIndices, uint64, error) {
-	return getBucketIndices(sr, addrKeyWithPrefix(addr, _voterIndex))
-}
-
-func putVoterBucketIndex(sm protocol.StateManager, addr address.Address, index uint64) error {
-	return putBucketIndex(sm, addrKeyWithPrefix(addr, _voterIndex), index)
+func (csr *candSR) GetVoterBucketIndices(addr address.Address) (*BucketIndices, uint64, error) {
```
Comment (CoderZhi) [inline action/protocol/staking/bucket_index.go:69]: rename  Public function`GetVoterBucketIndices` to `VoterBucketIndices`, same for the others

File: action/protocol/staking/candidate_statemanager.go [hunk 5]
```diff
@@ -76,6 +96,16 @@ func NewCandidateStateManager(sm protocol.StateManager, enableSMStorage bool) (C
 	return csm, nil
 }
 
+func smToCsm(sm protocol.StateManager) CandidateStateManager {
```
Comment (CoderZhi) [inline action/protocol/staking/candidate_statemanager.go:99]: rename `smToCsm` to `newCandidateStateManager`

File: action/protocol/staking/candidate_statereader.go [hunk 6]
```diff
@@ -43,6 +84,12 @@ type (
 	}
 )
 
+func srToCsr(sr protocol.StateReader) CandidateStateReader {
```
Comment (CoderZhi) [inline action/protocol/staking/candidate_statereader.go:87]: rename `srToCsr` to `newCandidateStateReader`

File: action/protocol/staking/candidate_statereader.go [hunk 7]
```diff
@@ -133,7 +180,7 @@ func CreateBaseView(sr protocol.StateReader, enableSMStorage bool) (*ViewData, u
 		return nil, 0, ErrMissingField
 	}
 
-	all, height, err := getAllCandidates(sr)
+	all, height, err := srToCsr(sr).getAllCandidates()
```
Comment (CoderZhi) [inline action/protocol/staking/candidate_statereader.go:183]:
```
csr := newCandidateStateReader(sr)
all, height, err := csr.getAllCandidates()
```

File: action/protocol/staking/candidate_statereader.go [hunk 8]
```diff
@@ -143,7 +190,7 @@ func CreateBaseView(sr protocol.StateReader, enableSMStorage bool) (*ViewData, u
 		return nil, height, err
 	}
 
-	pool, err := NewBucketPool(sr, enableSMStorage)
+	pool, err := srToCsr(sr).NewBucketPool(enableSMStorage)
```
Comment (CoderZhi) [inline action/protocol/staking/candidate_statereader.go:193]: reuse csr here

File: action/protocol/staking/candidate_statemanager.go [hunk 9]
```diff
@@ -136,18 +166,158 @@ func (csm *candSM) CreditBucketPool(amount *big.Int) error {
...
+	return err
+}
+
+func (csm *candSM) putBucketAndIndex(bucket *VoteBucket) (uint64, error) {
+	index, err := csm.putBucket(bucket)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to put bucket")
+	}
+
+	if err := csm.PutVoterBucketIndex(bucket.Owner, index); err != nil {
```
Comment (CoderZhi) [inline action/protocol/staking/candidate_statemanager.go:240]: Calling public function inside of a private function is not recommended, but let's leave it to next PR.

**Review by CoderZhi — APPROVED** [review]
much cleaner and clearer. Thanks for being patient.

**Review by Liuhaai — APPROVED** [review]

**Pull request merged**